src/storage/db/snapshot.rs

```diff
@@ -264,24 +296,79 @@ pub async fn upload_to_s3(
             .to_string();
         let key = format!("{}/{}", upload_dir, file_name);
 
-        info!(key, "Uploading chunk to s3");
-        let byte_stream = ByteStream::from_path(entry.path()).await?;
-        let upload_result = s3_client
-            .put_object()
-            .key(key.clone())
+        let file = tokio::fs::File::open(entry.path()).await?;
+        let mut reader = BufReader::new(file);
+
+        let create_res = s3_client
+            .create_multipart_upload()
             .bucket(snapshot_config.s3_bucket.clone())
-            .body(byte_stream)
+            .key(key.clone())
             .send()
-            .await;
-        if let Err(err) = &upload_result {
-            info!(
-                "Error uploading to s3: {}, key: {}, bucket: {}",
-                DisplayErrorContext(err),
-                key,
-                snapshot_config.s3_bucket
-            );
+            .await?;
+
+        let upload_id = create_res.upload_id().unwrap();
+        let mut parts = Vec::new();
+        // 5 MB is the minimum size
+        let mut buf = [0; UPLOAD_CHUNK_SIZE];
+        let mut part_number = 0;
+        while let Ok(bytes_read) = reader.read(&mut buf).await {
+            if bytes_read == 0 {
+                break;
+            }
+
+            part_number += 1;
+            info!(key, part_number, "Uploading snapshot chunk to s3");
+
+            match s3_client
+                .upload_part()
+                .bucket(snapshot_config.s3_bucket.clone())
+                .key(key.clone())
+                .upload_id(upload_id)
+                .part_number(part_number)
+                .body(ByteStream::from(buf.to_vec()))
+                .send()
+                .await
+            {
+                Err(err) => {
+                    info!(key, "Aborting multipart snapshot upload");
+                    s3_client
+                        .abort_multipart_upload()
+                        .bucket(snapshot_config.s3_bucket.clone())
+                        .key(key.clone())
+                        .upload_id(upload_id)
+                        .send()
+                        .await?;
+                    statsd_client.count_with_shard(shard_id, "snapshots.failed_upload", 1);
+                    return Err(SnapshotError::UploadPartError(err));
+                }
+                Ok(upload_res) => {
+                    parts.push(
+                        aws_sdk_s3::types::CompletedPart::builder()
+                            .part_number(part_number)
+                            .e_tag(upload_res.e_tag().unwrap())
+                            .build(),
+                    );
+                }
+            }
         }
-        upload_result?;
+
+        info!(key, "Finished uploading snapshot to s3");
+        statsd_client.count_with_shard(shard_id, "snapshots.successful_upload", 1);
```

Comment:
Let's emit on failure as well.

Reply:
There's a metric for failure above `snapshots.failed_upload`.